src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectFactory.cs

```diff
@@ -23,33 +28,50 @@ internal sealed class VisualStudioProjectFactory : IVsTypeScriptVisualStudioProj
         private readonly VisualStudioWorkspaceImpl _visualStudioWorkspaceImpl;
         private readonly HostDiagnosticUpdateSource _hostDiagnosticUpdateSource;
         private readonly ImmutableArray<Lazy<IDynamicFileInfoProvider, FileExtensionsMetadata>> _dynamicFileInfoProviders;
 
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
         public VisualStudioProjectFactory(
+            IThreadingContext threadingContext,
             VisualStudioWorkspaceImpl visualStudioWorkspaceImpl,
             [ImportMany] IEnumerable<Lazy<IDynamicFileInfoProvider, FileExtensionsMetadata>> fileInfoProviders,
             HostDiagnosticUpdateSource hostDiagnosticUpdateSource)
         {
+            _threadingContext = threadingContext;
             _visualStudioWorkspaceImpl = visualStudioWorkspaceImpl;
             _dynamicFileInfoProviders = fileInfoProviders.AsImmutableOrEmpty();
             _hostDiagnosticUpdateSource = hostDiagnosticUpdateSource;
         }
 
-        public VisualStudioProject CreateAndAddToWorkspace(string projectSystemName, string language)
-            => CreateAndAddToWorkspace(projectSystemName, language, new VisualStudioProjectCreationInfo());
+        public Task<VisualStudioProject> CreateAndAddToWorkspaceAsync(string projectSystemName, string language, CancellationToken cancellationToken)
+            => CreateAndAddToWorkspaceAsync(projectSystemName, language, new VisualStudioProjectCreationInfo(), cancellationToken);
 
-        public VisualStudioProject CreateAndAddToWorkspace(string projectSystemName, string language, VisualStudioProjectCreationInfo creationInfo)
+        public async Task<VisualStudioProject> CreateAndAddToWorkspaceAsync(
+            string projectSystemName, string language, VisualStudioProjectCreationInfo creationInfo, CancellationToken cancellationToken)
         {
-            // HACK: Fetch this service to ensure it's still created on the UI thread; once this is moved off we'll need to fix up it's constructor to be free-threaded.
+            // HACK: Fetch this service to ensure it's still created on the UI thread; once this is
+            // moved off we'll need to fix up it's constructor to be free-threaded.
+
+            // Since we're on the UI thread here anyways, use that as an opportunity to grab the
+            // IVsSolution object and solution file path.
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
             _visualStudioWorkspaceImpl.Services.GetRequiredService<VisualStudioMetadataReferenceManager>();
 
-            // HACK: since we're on the UI thread, ensure we initialize our options provider which depends on a UI-affinitized experimentation service
-            _visualStudioWorkspaceImpl.EnsureDocumentOptionProvidersInitialized();
+            var solution = (IVsSolution2)Shell.ServiceProvider.GlobalProvider.GetService(typeof(SVsSolution));
```

Comment:
💡 Use `IAsyncServiceProvider.GetServiceAsync`

Reply:
i don't really understand the relation between iAsyncServiceProvider, and ServiceProvider.GlobalProvider.

Reply:
done.